charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,259 @@
+            labels=create_charm_default_labels(
+                application_name=self.app.name,
+                model_name=self.model.name,
+                scope=resources["scope"],
+            ),
+            logger=self.logger,
         )
 
-        self.model.unit.status = ActiveStatus()
+    def _get_gateway_data(self) -> Tuple[str, str]:
+        """Retrieve gateway namespace and name from relation data."""
+        try:
+            gateway_data = self.gateway.get_relation_data()
+        except GatewayRelationError:
+            raise ErrorWithStatus("Waiting for gateway info relation", WaitingStatus)
+
+        return gateway_data["gateway_namespace"], gateway_data["gateway_name"]
 
-    def _check_leader(self):
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Apply K8s resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8s resources")
+        try:
+            for handler in {self.rbac_resource_handler, self.crd_resource_handler}:
+                handler.apply(force=force_conflicts)
+        except Exception as error:
+            raise GenericCharmRuntimeError("K8s resources creation failed", error)
+
+    def _check_leader(self) -> None:
+        """Check whether a unit is the leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
+
+    def _on_install(self, _) -> None:
+        """Handle install event."""
+        # deploy K8s resources to speed up deployment
+        self._apply_k8s_resources()
+
+    def _on_upgrade(self, _) -> None:
+        """Handle upgrade event."""
+        # force conflict resolution in K8s resources update
+        self._on_event(_, force_conflicts=True)
 
-    def _check_image_details(self):
+    def _on_remove(self, _) -> None:
+        """Handle remove event."""
+        delete_errors = []
+        self.unit.status = MaintenanceStatus("Removing K8s resources")
+        for handler in {self.rbac_resource_handler, self.crd_resource_handler}:
+            try:
+                handler.delete()
+            except ApiError as error:
+                delete_errors.append(error)
+
+        if delete_errors:
+            raise GenericCharmRuntimeError("Removing K8s resources failed", delete_errors)
```

Comment:
on remove, we do not want to re-raise 404 errors (eg: errors saying we tried to delete objects that no longer existed) since sometimes race conditions exist on cleanup.  Instead we just log them but then suppress the error.  For unknown errors, we reraise.

[example](https://github.com/canonical/seldon-core-operator/blob/58a245b049ea907eb8b05a924e9de98aed2f5e1b/src/charm.py#L409)

Reply:
Indeed! I had opened a PR to move this functionality to `chisme` but hadn't noticed that I'm blocking it:
* https://github.com/canonical/charmed-kubeflow-chisme/pull/62

Reply:
Let's merge the current PR and then fix this up in a day or two when the chisme PR merges